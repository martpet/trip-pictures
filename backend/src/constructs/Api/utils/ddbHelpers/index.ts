export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
