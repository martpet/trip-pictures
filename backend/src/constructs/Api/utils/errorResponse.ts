import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { ApiErrorResponseBody } from '~/types';

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

export const errorResponse = (
  traceId: string,
  options: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  } = options;

  const { envName } = globalLambdaProps;

  const errObj: ApiErrorResponseBody['error'] = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errObj.description = description;
  }

  if (error && (exposeError ?? envName !== 'production')) {
    errObj.error = String(error);
  }

  const body: ApiErrorResponseBody = {
    error: errObj,
  };

  console.error({ traceId, error });

  return {
    statusCode,
    body: JSON.stringify(body),
  };
};
