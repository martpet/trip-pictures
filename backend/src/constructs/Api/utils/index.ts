export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
