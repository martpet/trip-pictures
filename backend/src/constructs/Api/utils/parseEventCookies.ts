import { APIGatewayProxyEventV2 } from 'aws-lambda';
import cookie from 'cookie';

import { Cookies, PathWithCookies } from '../types';

export const parseEventCookies = <T extends PathWithCookies>(
  event: APIGatewayProxyEventV2
): Cookies<T> => {
  try {
    const cookieHeaderString = event.cookies?.join(';');
    if (!cookieHeaderString) throw new Error();
    return cookie.parse(cookieHeaderString);
  } catch (e) {
    return {};
  }
};
