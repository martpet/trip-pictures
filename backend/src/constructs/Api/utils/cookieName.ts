import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
