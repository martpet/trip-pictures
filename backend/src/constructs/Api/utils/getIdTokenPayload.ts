import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = async (idToken: string): Promise<IdTokenPayload> => {
  const camelcaseKeys = (await import('camelcase-keys')).default;
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
