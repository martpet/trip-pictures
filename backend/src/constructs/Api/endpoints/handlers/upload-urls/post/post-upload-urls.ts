import {
  APIGatewayProxyHandlerV2,
  createPhotoS3Key,
  createPresignedPost,
  EnvVars,
  errorResponse,
  getIdTokenPayload,
  maxPhotoUploadSize,
  PhotosTableItem,
  photosTableOptions,
  PostUploadUrlsRequest,
  PostUploadUrlsResponse,
  RouteHeaders,
  S3Client,
  StatusCodes,
} from 'lambda-layer';

import { ddbBatchGet } from '~/constructs/Api/utils';
import { IS_PHOTOS_TABLE_ITEM_CREATED_S3_TAG } from '~/constructs/Photos/consts';

const s3Client = new S3Client({});

const TAGGING = `<Tagging><TagSet><Tag><Key>${IS_PHOTOS_TABLE_ITEM_CREATED_S3_TAG}</Key><Value>false</Value></Tag></TagSet></Tagging>`;

export const handler: APIGatewayProxyHandlerV2<PostUploadUrlsResponse> = async (
  event
) => {
  const { authorization } = event.headers as RouteHeaders<'/settings'>;
  const { photoBucket } = process.env as EnvVars<'/upload-urls', 'POST'>;

  if (!authorization) {
    return errorResponse('Vf5Ph6qN1S');
  }

  if (!photoBucket) {
    return errorResponse('1rQkj3kpp4');
  }

  if (!event.body) {
    return errorResponse('G-luuHqI0s', { statusCode: StatusCodes.BAD_REQUEST });
  }

  let requestData;

  try {
    requestData = JSON.parse(event.body) as PostUploadUrlsRequest;
  } catch (error) {
    return errorResponse('9210145fdf', { statusCode: StatusCodes.BAD_REQUEST, error });
  }

  const existingItemsInDb = await ddbBatchGet<PhotosTableItem>({
    pkValues: requestData.map(({ fingerprint }) => fingerprint),
    tableOptions: photosTableOptions,
  });

  const existingFingerprintsInDb = existingItemsInDb.map(
    ({ fingerprint }) => fingerprint
  );

  const uniqueRequestData = requestData.filter(
    ({ fingerprint }) => !existingFingerprintsInDb.includes(fingerprint)
  );

  const { sub: userId } = await getIdTokenPayload(authorization);

  const uploadUrlsEntries = await Promise.all(
    uniqueRequestData.map(async ({ id, fingerprint, digest }) => {
      const key = createPhotoS3Key({ userId, fingerprint });
      const presignedPost = await createPresignedPost(s3Client, {
        Bucket: photoBucket,
        Key: key,
        Expires: 600,
        Fields: {
          'x-amz-checksum-algorithm': 'SHA256',
          'x-amz-checksum-sha256': digest,
          tagging: TAGGING,
        },
        Conditions: [
          ['content-length-range', 0, maxPhotoUploadSize],
          { tagging: TAGGING },
        ],
      });
      return [id, presignedPost];
    })
  );

  return {
    presignedPosts: Object.fromEntries(uploadUrlsEntries),
    existingFingerprintsInDb,
  };
};
