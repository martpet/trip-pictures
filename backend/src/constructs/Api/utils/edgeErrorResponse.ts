import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
