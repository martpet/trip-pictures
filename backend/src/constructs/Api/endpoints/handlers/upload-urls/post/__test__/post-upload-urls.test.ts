import { createPresignedPost } from '@aws-sdk/s3-presigned-post';

import {
  ddbBatchGet,
  itCalls,
  itGetsIdToken,
  itHasEnvVars,
  itHasJsonBody,
  itResolves,
} from '~/constructs/Api/utils';
import { createPhotoS3Key } from '~/utils';

import { handler } from '../post-upload-urls';

vi.mock('@aws-sdk/s3-presigned-post');
vi.mock('@aws-sdk/client-s3');
vi.mock('~/utils/createPhotoS3Key');
vi.mock('~/constructs/Api/utils/errorResponse');
vi.mock('~/constructs/Api/utils/ddbHelpers/ddbBatchGet');
vi.mock('~/constructs/Api/utils/getIdTokenPayload');
vi.mock('~/constructs/Api/utils/ddbHelpers/ddbBatchGet');

process.env.photoBucket = 'dummyPhotoBucket';

beforeEach(() => {
  vi.mocked(ddbBatchGet).mockResolvedValue([{ fingerprint: 'dummyFingerprint1' }]);
});

const args = [
  {
    headers: { authorization: 'dummyAuthorizationHeader' },
    body: JSON.stringify([
      {
        id: 'dummyId1',
        fingerprint: 'dummyFingerprint1',
        hash: 'dummyHash1',
      },
      {
        id: 'dummyId2',
        fingerprint: 'dummyFingerprint2',
        hash: 'dummyHash2',
      },
    ]),
  },
] as unknown as Parameters<typeof handler>;

describe('post-upload-urls', () => {
  itHasJsonBody(handler, args);
  itHasEnvVars(['photoBucket'], handler, args);
  itGetsIdToken(handler, args);
  itCalls(ddbBatchGet, handler, args);
  itCalls(createPresignedPost, handler, args);
  itCalls(createPhotoS3Key, handler, args);
  itResolves(handler, args);
});
