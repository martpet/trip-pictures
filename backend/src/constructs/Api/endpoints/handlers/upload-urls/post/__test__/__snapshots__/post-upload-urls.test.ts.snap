// Vitest Snapshot v1

exports[`post-upload-urls > calls "createPhotoS3Key" with correct args 1`] = `
[
  [
    {
      "fingerprint": "dummyFingerprint2",
      "userId": "dummySub",
    },
  ],
]
`;

exports[`post-upload-urls > calls "createPresignedPost" with correct args 1`] = `
[
  [
    DummyS3Client {},
    {
      "Bucket": "dummyPhotoBucket",
      "Conditions": [
        [
          "content-length-range",
          0,
          15728640,
        ],
        {
          "tagging": "<Tagging><TagSet><Tag><Key>isTableItemCreated</Key><Value>false</Value></Tag></TagSet></Tagging>",
        },
      ],
      "Expires": 600,
      "Fields": {
        "tagging": "<Tagging><TagSet><Tag><Key>isTableItemCreated</Key><Value>false</Value></Tag></TagSet></Tagging>",
        "x-amz-checksum-algorithm": "SHA256",
        "x-amz-checksum-sha256": undefined,
      },
      "Key": "dummyPhotoS3Key",
    },
  ],
]
`;

exports[`post-upload-urls > calls "ddbBatchGet" with correct args 1`] = `
[
  [
    {
      "pkValues": [
        "dummyFingerprint1",
        "dummyFingerprint2",
      ],
      "tableOptions": {
        "partitionKey": {
          "name": "fingerprint",
          "type": "S",
        },
        "stream": "NEW_IMAGE",
        "tableName": "TP-photos",
      },
    },
  ],
]
`;

exports[`post-upload-urls > calls "getIdTokenPayload" with correct args 1`] = `
[
  [
    "dummyAuthorizationHeader",
  ],
]
`;

exports[`post-upload-urls > resolves with a correct value 1`] = `
{
  "existingFingerprintsInDb": [
    "dummyFingerprint1",
  ],
  "presignedPosts": {
    "dummyId2": "dummyPresignedPost",
  },
}
`;

exports[`post-upload-urls > when "authorization" header is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "Vf5Ph6qN1S",
  ],
]
`;

exports[`post-upload-urls > when "authorization" header is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "event.body" is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "G-luuHqI0s",
    {
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-upload-urls > when "event.body" is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "event.body" is not JSON > calls "errorResponse" with correct args 1`] = `
[
  [
    "9210145fdf",
    {
      "error": [SyntaxError: Unexpected token o in JSON at position 1],
      "statusCode": 400,
    },
  ],
]
`;

exports[`post-upload-urls > when "event.body" is not JSON > resolves with a correct value 1`] = `"dummyErrorResponse"`;

exports[`post-upload-urls > when "photoBucket" env var is missing > calls "errorResponse" with correct args 1`] = `
[
  [
    "1rQkj3kpp4",
  ],
]
`;

exports[`post-upload-urls > when "photoBucket" env var is missing > resolves with a correct value 1`] = `"dummyErrorResponse"`;
